parquet-column/src/main/java/org/apache/parquet/column/values/bloom/Bloom.java

```diff
@@ -0,0 +1,423 @@
+
+    // Add bloom filter algorithm
+    inputs.add(BytesInput.fromInt(this.algorithm.ordinal()));
+
+    // Add bitset.
+    inputs.add(BytesInput.from(bitset, 0, bitset.length));
+
+    return BytesInput.concat(inputs);
+  }
+
+  private int[] setMask(int key) {
+    // The block based algorithm needs 8 odd SALT values to calculate index
+    // of bit to set in block.
+    final int SALT[] = {0x47b6137b, 0x44974d91, 0x8824ad5b, 0xa2b7289d,
+      0x705495c7, 0x2df1424b, 0x9efc4947, 0x5c6bfb31};
+
+    int mask[] = new int[8];
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = key * SALT[i];
+    }
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = mask[i] >>> 27;
+    }
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = 0x1 << mask[i];
+    }
+
+    return mask;
+  }
+
+  /**
+   * Add an element to bloom filter, the element content is represented by
+   * the hash value of its plain encoding result.
+   * @param hash hash result of element.
+   */
+  private void addElement(long hash) {
+    int bucketIndex = (int)(hash >> 32) & (bitset.length / BYTES_PER_BUCKET - 1);
+    int key = (int)hash;
+
+    // Calculate bit mask for bucket.
+    int mask[] = setMask(key);
+
+    for (int i = 0; i < 8; i++) {
+      int bitsetIndex = bucketIndex * BYTES_PER_BUCKET + i * 4;
+      bitset[bitsetIndex] |= (byte)(mask[i] >>> 24);
+      bitset[bitsetIndex + 1] |= (byte)(mask[i] >>> 16);
+      bitset[bitsetIndex + 2] |= (byte)(mask[i] >>> 8);
+      bitset[bitsetIndex + 3] |= (byte)(mask[i]);
+    }
+  }
+
+  /**
+   * Determine where an element is in set or not.
+   * @param hash the hash value of element plain encoding result.
+   * @return false if element is must not in set, true if element probably in set.
+   */
+  private boolean isContain(long hash) {
+    int bucketIndex = (int)(hash >> 32) & (bitset.length / BYTES_PER_BUCKET - 1);
+    int key = (int)hash;
+
+    // Calculate bit mask for bucket.
+    int mask[] = setMask(key);
+
+    for (int i = 0; i < 8; i++) {
+      byte set = 0;
+      int bitsetIndex = bucketIndex * BYTES_PER_BUCKET + i * 4;
+      set |= bitset[bitsetIndex] & ((byte)(mask[i] >>> 24));
+      set |= bitset[bitsetIndex + 1] & ((byte)(mask[i] >>> 16));
+      set |= bitset[bitsetIndex + 2] & ((byte)(mask[i] >>> 8));
+      set |= bitset[bitsetIndex + 3] & ((byte)mask[i]);
+      if (0 == set) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Calculate optimal size according to the number of distinct values and false positive probability.
+   * @param n: The number of distinct values.
+   * @param p: The false positive probability.
+   * @return optimal number of bits of given n and p.
+   */
+  public static int optimalNumOfBits(long n, double p) {
+    Preconditions.checkArgument((p > 0.0 && p < 1.0),
+      "FPP should be less than 1.0 and great than 0.0");
+
+    final double M = -8 * n / Math.log(1 - Math.pow(p, 1.0 / 8));
```

Comment:
Given 0<p<1 => 0 < pow(p, 0.125) < 1 => log (1- pow(p, 0.125)) < 0 => M>0 or (M<0 overflow). 
 
So M<0 when result overflow according to formula, it should be set to predefined maximum number of bytes ParquetProperties.DEFAULT_MAXIMUM_BLOOM_FILTER_SIZE << 3.